### PR TITLE
Resize RenderWidget properly when using different DPI scaling.

### PR DESCRIPTION
**Pull request by rlober**
See #1262 for details.

### CODE REVIEW

File: src/rviz/ogre_helpers/render_widget.cpp [hunk 1]
```diff
@@ -81,12 +81,12 @@ RenderWidget::RenderWidget( RenderSystem* render_system, QWidget *parent )
 
 #if QT_VERSION < QT_VERSION_CHECK(5, 0, 0)
   QApplication::syncX();
-  double pixel_ratio = 1.0;
```
Comment (rhaschke) [inline src/rviz/ogre_helpers/render_widget.cpp]: This variable should stay double, shouldn't it?

File: src/rviz/ogre_helpers/render_widget.h [hunk 2]
```diff
@@ -62,6 +62,8 @@ class RenderWidget: public QWidget
   Ogre::RenderWindow* render_window_;
 
   QFrame* renderFrame;
+
+  int pixel_ratio_;
```
Comment (VictorLamoine) [inline src/rviz/ogre_helpers/render_widget.h]: I would initialize the variable to 1 here, then you can remove the assignment when Qt < 5 and you are always safe if someone does something wrong with the pre-processor macros.

**Comment by atyshka** [comment]
+1 on this. I have a HiDPI mac and have this issue with RViz2 in ROS. It would be great if this was merged and added to the ros 2 repo also.

File: src/rviz/ogre_helpers/render_widget.h (hunk 2 above)
Comment (sippeyxp) [reply src/rviz/ogre_helpers/render_widget.h]: Please keep the variable double/float. AFAIK, pixel ratio does not have to be integers. For example, setting "export QT_SCALE_FACTOR=2.5" makes QWindow::devicePixelRatio() returns 2.5.

File: src/rviz/ogre_helpers/render_widget.cpp (hunk 1 above)
Comment (rlober) [reply src/rviz/ogre_helpers/render_widget.cpp]: Sorry for the late reply - I haven't checked on this PR in almost a year... Yes it should be a `double`. Not sure why I made it an `int` at the time :thinking: .

Comment (rlober) [reply src/rviz/ogre_helpers/render_widget.cpp]: Just recompiled with `double` and realized why...

```
 error: invalid operands of types ‘double’ and ‘int’ to binary ‘operator%’
     render_window_->resize( width()*pixel_ratio_ + (width()*pixel_ratio_ % 2), height()*pixel_ratio_ );
```
It's easily fixed with the following:
```
render_window_->resize( width()*pixel_ratio_ + (static_cast<int>(width()*pixel_ratio_) % 2), height()*pixel_ratio_ );
```
But I haven't tested yet on my mac to see if everything looks good. I will try to do this later in the week.

Comment (rlober) [reply src/rviz/ogre_helpers/render_widget.cpp]: Checked the new changes on my mac and everything seems to be working properly.

File: src/rviz/ogre_helpers/render_widget.h (hunk 2 above)
Comment (rlober) [reply src/rviz/ogre_helpers/render_widget.h]: Fixed.

**Comment by rlober** [comment]
So all the requested changes have been made and everything seems to work as expected. I can't help the ABI change without making a global variable.

**Comment by atyshka** [comment]
Any progress on merging this?

**Comment by tompe17** [comment]
I just manually applied tese patches. I really need them...  So a merge would be very good.

But I noticed when using these patches together with jsk-visualization and overlay-menu that the menu was not placed in the middle.  The rest of the things seems to work.

**Comment by rhaschke** [comment]
I would really like to validate those changes before merging. Could anybody (maybe @sippeyxp) hint me at instructions on how to set up a high-dpi monitor on Linux (Ubuntu) to 
a) reproduce the issue and 
b) validate that this PR fixes it.

@tompe17, could you please elaborate more on the issue described in https://github.com/ros-visualization/rviz/pull/1263#issuecomment-515191209?

**Comment by rlober** [comment]
@rhaschke what kind of monitors do you have access to?

**Comment by rhaschke** [comment]
I just have a laptop (1920x1080) with an external monitor (1920x1200). But I guess, by configuring it somehow, I can pretend that the monitor is high-dpi (resulting in huge fonts / gui elements on that screen), right? So my questions is: How do I configure it like this?

**Comment by atyshka** [comment]
Should be able to do it in Ubuntu display settings... unless it knows the resolution and only offers scaling for hidpi. In that case maybe you could try a vm with custom display settings?

**Comment by rhaschke** [comment]
I don't get high-dpi display settings offered.

**Comment by VictorLamoine** [comment]
@rhaschke https://askubuntu.com/a/960041/73160 does this help ?

**Comment by sippeyxp** [comment]
Verified the patch fixed issue in my build (lunar). Builds fine and looks perfectly on 3840x2160 15" display.

Not sure why the cont. integration is not happy.

**Comment by rlober** [comment]
@rhaschke 

I don't know if you can trick the high DPI checking in QT without having the physical pixels. Let me know if @VictorLamoine 's link works for you. 

@sippeyxp 

Thanks for checking. The CI is unhappy due to the change to the header `render_widget.h` ([see here](https://github.com/ros-visualization/rviz/pull/1263/commits/0d9d5b5eff6fa0215cac26d2b183dcab44dd8608#diff-c61e575a644d99fb12f1a08cabb1bd47R66)) causing an ABI incompatibility with `librviz` ([see here](https://travis-ci.org/ros-visualization/rviz/builds/571887316#L2305)).

**Comment by tompe17** [comment]
The issue I saw was that an overlay menu that is centered on low resolution devices seemed to be centered in the top left 1/4 of the high resolution screen. But thing still works so do not let this block the patch. The alternative is that it does not work at all.

**Comment by rhaschke** [comment]
I managed to have a custom pixel ratio via:
`QT_SCALE_FACTOR=1.5 rviz`

**Comment by rlober** [comment]
Can this be merged into melodic-devel also ?

**Comment by rhaschke** [comment]
It is already forward-ported (1342d2e17b447a24e45a1fd34380a9dda4848421) and released into Melodic. Are there any open issues in Melodic?

**Comment by Tacoma** [comment]
Hi, I hope this reaches the right people :D
I noticed that the QApplication::sync() that was added in line 84 of the render_widget.cpp causes the VisualizationFrame to be initialized twice. I wonder if this is expected behaviour and why this sync() call is needed.
I noticed the problem because I advertise a service in a custom rviz display plugin. This fails when starting rviz with a config including this plugin.
Is it save for me to remove the QApplication::sync() call in my rviz version? I'm fine with building rviz myself.

I printed the stack-trace every time my plugin is initialized to find the problem:
```
[ INFO] Initializing preview service.
[ INFO] waitForService: Service [/save_preview] has not been advertised, waiting...
 0# TextureDisplay::onInitialize()
 1# rviz::Display::initialize(rviz::DisplayContext*) in librviz.so
 2# rviz::DisplayGroup::load(rviz::Config const&) in librviz.so
 3# rviz::VisualizationManager::load(rviz::Config const&) in librviz.so
 4# rviz::VisualizationFrame::load(rviz::Config const&) in librviz.so
 5# rviz::VisualizationFrame::loadDisplayConfig(QString const&) in librviz.so
 6# rviz::VisualizationFrame::initialize(QString const&) in librviz.so
 7# RvizPluginFactory::create(QString const&, QUrl const&, QStringList const&, QStringList const&) const
 8# 0x00007F219C593861 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
 9# 0x00007F219D5A51D4 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
10# 0x00007F219D5A66BA in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
11# 0x00007F219D154FF8 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
12# 0x00007F219D15A2F4 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
13# 0x00007F219D62733C in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
14# 0x00007F219D630830 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
15# 0x00007F219D63087C in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
16# 0x00007F219D6D496E in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
17# QObject::event(QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
18# QApplicationPrivate::notify_helper(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
19# QApplication::notify(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
20# QCoreApplication::notifyInternal2(QObject*, QEvent*) in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
21# QTimerInfoList::activateTimers() in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
22# 0x00007F219AC8B391 in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
23# g_main_context_dispatch in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
24# 0x00007F218F1DD650 in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
25# g_main_context_iteration in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
26# QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) in /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
27# QGuiApplication::sync() in /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
28# rviz::RenderWidget::RenderWidget(rviz::RenderSystem*, QWidget*) in librviz.so
29# rviz::QtOgreRenderWindow::QtOgreRenderWindow(QWidget*) in librviz.so
30# rviz::RenderPanel::RenderPanel(QWidget*) in librviz.so
31# rviz::VisualizationFrame::initialize(QString const&) in librviz.so
32# RvizPluginFactory::create(QString const&, QUrl const&, QStringList const&, QStringList const&) const
33# 0x00007F219C593861 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
34# 0x00007F219D5A51D4 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
35# 0x00007F219D5A66BA in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
36# 0x00007F219D154FF8 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
37# 0x00007F219D15A2F4 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
38# 0x00007F219D7B3258 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
39# 0x00007F219D27D98E in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
40# 0x00007F219D29681A in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
41# 0x00007F219D2A5DAC in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
42# 0x00007F219D124CA6 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
43# 0x00007F219D1246F1 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
44# 0x00007F219D3AC813 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
45# 0x00007F219D3B824A in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
46# 0x00007F219D3B838E in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
47# 0x00007F219DFB9C3F in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
48# 0x00007F219CAEB568 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
49# 0x00007F219C8A399D in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
[ INFO] service advertised.
...
[ INFO] Initializing preview service.
 0# TextureDisplay::onInitialize()
 1# rviz::Display::initialize(rviz::DisplayContext*) in librviz.so
 2# rviz::DisplayGroup::load(rviz::Config const&) in librviz.so
 3# rviz::VisualizationManager::load(rviz::Config const&) in librviz.so
 4# rviz::VisualizationFrame::load(rviz::Config const&) in librviz.so
 5# rviz::VisualizationFrame::loadDisplayConfig(QString const&) in librviz.so
 6# rviz::VisualizationFrame::initialize(QString const&) in librviz.so
 7# RvizPluginFactory::create(QString const&, QUrl const&, QStringList const&, QStringList const&) const
 8# 0x00007F219C593861 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
 9# 0x00007F219D5A51D4 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
10# 0x00007F219D5A66BA in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
11# 0x00007F219D154FF8 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
12# 0x00007F219D15A2F4 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
13# 0x00007F219D7B3258 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
14# 0x00007F219D27D98E in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
15# 0x00007F219D29681A in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
16# 0x00007F219D2A5DAC in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
17# 0x00007F219D124CA6 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
18# 0x00007F219D1246F1 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
19# 0x00007F219D3AC813 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
20# 0x00007F219D3B824A in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
21# 0x00007F219D3B838E in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
22# 0x00007F219DFB9C3F in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
23# 0x00007F219CAEB568 in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
24# 0x00007F219C8A399D in /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5
[ERROR] Tried to advertise a service that is already advertised in this node [/save_preview]
```

**Comment by rhaschke** [comment]
Sorry @Tacoma, but I can't reproduce your issue. For me the stack trace looks like this:
```
#0  0x00007fff98ae80e1 in rviz::GridDisplay::onInitialize() (this=0x55555648cdc0) at /vol/sandbox/rviz/src/rviz/src/rviz/default_plugin/grid_display.cpp:115
#1  0x00007ffff7a6c48d in rviz::Display::initialize(rviz::DisplayContext*) (this=this@entry=0x55555648cdc0, context=<optimized out>) at /vol/sandbox/rviz/src/rviz/src/rviz/display.cpp:95
#2  0x00007ffff7a83d01 in rviz::DisplayGroup::load(rviz::Config const&) (this=0x555556087380, config=...) at /vol/sandbox/rviz/src/rviz/src/rviz/display_group.cpp:109
#3  0x00007ffff7b48a70 in rviz::VisualizationManager::load(rviz::Config const&) (this=this@entry=0x555556660ba0, config=...) at /vol/sandbox/rviz/src/rviz/src/rviz/visualization_manager.cpp:510
#4  0x00007ffff7b441bb in rviz::VisualizationFrame::load(rviz::Config const&) (this=0x55555586bad0, config=...) at /vol/sandbox/rviz/src/rviz/src/rviz/visualization_frame.cpp:850
#5  0x00007ffff7b44f9f in rviz::VisualizationFrame::loadDisplayConfigHelper(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (this=this@entry=0x55555586bad0, full_path="/tmp/empty.rviz") at /vol/sandbox/rviz/src/rviz/src/rviz/visualization_frame.cpp:765
#6  0x00007ffff7b4514c in rviz::VisualizationFrame::loadDisplayConfig(QString const&) (this=this@entry=0x55555586bad0, qpath=...) at /vol/sandbox/rviz/src/rviz/src/rviz/visualization_frame.cpp:736
#7  0x00007ffff7b45b5b in rviz::VisualizationFrame::initialize(QString const&) (this=this@entry=0x55555586bad0, display_config_file=...) at /vol/sandbox/rviz/src/rviz/src/rviz/visualization_frame.cpp:353
#8  0x00007ffff7b4f99e in rviz::VisualizerApp::init(int, char**) (this=0x7fffffffd430, argc=<optimized out>, argv=<optimized out>) at /vol/sandbox/rviz/src/rviz/src/rviz/visualizer_app.cpp:212
#9  0x0000555555554ec1 in main(int, char**) (argc=3, argv=0x7fffffffd598) at /vol/sandbox/rviz/src/rviz/src/rviz/main.cpp:40
```
Looks like you are creating a new VisualizationFrame in a plugin?
In any case, please open a new issue and refer to related ones if you see fit.

**Pull request merged**